### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by georgekuruvillak**
**What this PR does / why we need it**:
This PR adds unit tests for testing dependency-watchdog. 
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: pkg/restarter/restarter_test.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	}
+	go func() {
+		pl, err := f.client.CoreV1().Pods(metav1.NamespaceDefault).List(metav1.ListOptions{})
+		if err != nil {
+			t.Fatalf("error fetching pods: %v", err)
+		}
+		before := len(pl.Items)
+		t.Logf("Number of pods before the watchdog started: %v.", before)
+		watcher.Add(pC)
+		watcher.Add(pH)
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go]: Why not use the already existing mechanism in MCM tests? Even dispatching initial objects has been [implemented](https://github.com/gardener/machine-controller-manager/blob/master/pkg/fakeclient/client.go#L297:19).

File: pkg/restarter/restarter_test.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	// pC := newPodInCrashloop("pod-0", map[string]string{
+	// 	"garden.sapcloud.io/role": "controlplane",
+	// 	"role":                    "NotEtcd",
+	// })
+	pH := newPodHealthy("pod-1", map[string]string{
+		"garden.sapcloud.io/role": "controlplane",
+		"role":                    "NotEtcd",
+	})
+
+	f.eLister = append(f.eLister, e)
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go]:
```suggestion
	f.endpoints = append(f.endpoints, e)
```

File: pkg/restarter/restarter_test.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	//f.pLister = append(f.pLister, p)
+	f.objects = append(f.objects, e, pH)
+	f.client = fake.NewSimpleClientset(f.objects...)
+	watcher := watch.NewFake()
+	f.client.PrependWatchReactor("pods", test.DefaultWatchReactor(watcher, nil))
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go]: Why not use the already existing mechanism in MCM tests? Even dispatching initial objects has been [implemented](https://github.com/gardener/machine-controller-manager/blob/master/pkg/fakeclient/client.go#L297:19).

File: pkg/restarter/restarter_test.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+		t.Fatalf("error creating Deployment controller: %v", err)
+	}
+	go func() {
+		pl, err := f.client.CoreV1().Pods(metav1.NamespaceDefault).List(metav1.ListOptions{})
+		if err != nil {
+			t.Fatalf("error fetching pods: %v", err)
+		}
+		before := len(pl.Items)
+		t.Logf("Number of pods before the watchdog started: %v.", before)
+		watcher.Add(pH)
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go]: Why not use the already existing mechanism in MCM tests? Even dispatching initial objects has been [implemented](https://github.com/gardener/machine-controller-manager/blob/master/pkg/fakeclient/client.go#L297:19).

File: pkg/restarter/restarter_test.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+		if err != nil {
+			t.Fatalf("error fetching pods: %v", err)
+		}
+
+		t.Logf("Making pod go into CrashloopBackoff and wait for 2 seconds.")
+		pU, err := f.client.CoreV1().Pods(metav1.NamespaceDefault).Update(makePodUnhealthy(pH))
+		if err != nil {
+			t.Fatalf("error updating pods: %v", err)
+		}
+		watcher.Modify(pU)
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go]: Why not use the already existing mechanism in MCM tests? Even dispatching initial objects has been [implemented](https://github.com/gardener/machine-controller-manager/blob/master/pkg/fakeclient/client.go#L297:19).

File: pkg/restarter/restarter_test.go.tmp [hunk 6]
```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
```
Comment (amshuman-kr) [inline pkg/restarter/restarter_test.go.tmp]: Do we need this file?

File: pkg/restarter/restarter_test.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+	client *fake.Clientset
+	// Objects to put in the store.
+	eLister []*v1.Endpoints
+	pLister []*v1.Pod
+
+	// Actions expected to happen on the client. Objects from here are also
+	// preloaded into NewSimpleFake.
+	actions []test.Action
+	objects []runtime.Object
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply pkg/restarter/restarter_test.go]: In the current implementation the informers are populated from the respective arrays. With object array, we need to type check and populate. I tried using the watchable object tracker. Some glitch preventing the watch from triggering. So I reverted and thought to let these tests go in for now and raise another issue for implementing using the watchable tracker. WDYT?

File: pkg/restarter/restarter_test.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+		f.client,
+		controller.NoResyncPeriodFunc(),
+		informers.WithNamespace(deps.Namespace),
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = labelSelector.String()
+		}))
+
+	c := NewController(f.client, informers, deps, watchDuration, stopCh)
+	for _, d := range f.eLister {
+		informers.Apps().V1().Deployments().Informer().GetIndexer().Add(d)
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply pkg/restarter/restarter_test.go]: When I tried a Create it was going into a long wait? Create should ideally populate the informer. I took inspiration from [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller_test.go#L183).

File: pkg/restarter/restarter_test.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+	}
+	for _, pod := range f.pLister {
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply pkg/restarter/restarter_test.go]: When I tried a Create it was going into a long wait? Create should ideally populate the informer. I took inspiration from here.

File: pkg/restarter/restarter_test.go [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+		informers.Core().V1().Pods().Informer().GetIndexer().Add(pod)
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply pkg/restarter/restarter_test.go]: When I tried a Create it was going into a long wait? Create should ideally populate the informer. I took inspiration from here.

File: pkg/restarter/restarter_test.go (hunk 5 above)
Comment (georgekuruvillak) [reply pkg/restarter/restarter_test.go]: I tried using the watchable object tracker. Some glitch preventing the watch from triggering. So I reverted and thought to let these tests go in for now and raise another issue for implementing using the watchable tracker. WDYT?

**Pull request merged**